lib/Parse/ParseDecl.cpp

```diff
@@ -2406,10 +2406,35 @@ Decl *Parser::ParseDeclarationAfterDeclaratorAndAttributes(
           StopTokens.push_back(tok::r_paren);
         SkipUntil(StopTokens, StopAtSemi | StopBeforeMatch);
         Actions.ActOnInitializerError(ThisDecl);
-      } else
-        Actions.AddInitializerToDecl(ThisDecl, Init.get(),
+      } else {
+        auto *InitExpr = Init.get(); // This won't fail, since isInvalid() was checked above.
+        if (D.getDeclSpec().isUnpackSpecified()) {
+          auto TypeVars = D.getDeclSpec().typeVariables();
+          if (TypeVars.size() != 1) {
+            Diag(D.getBeginLoc(), diag::err_unpack_expected_one_type_variable) << static_cast<unsigned int>(TypeVars.size());
+            return nullptr;
+          }
+          auto TVar = Actions.getASTContext().getTypedefType(TypeVars[0], QualType());
+          auto *SourceInfo = Actions.Context.getTrivialTypeSourceInfo(TVar, TypeVars[0]->getLocation());
+          auto TypeArg = TypeArgument { TVar, SourceInfo };
+          // The initializer must have an existential type.
+          // e.g. _Unpack (T) struct Foo<T> fooAbs = expr // expr must have type '_Exists(A, struct Foo<A>)'
+          auto InitExprType = InitExpr->getType().getCanonicalType().getTypePtr();
```

Comment:
You should use the getType->getAs<ExistentialType>() instead.  In general, if you are manipulating CanonicalTypes directly in semantic checking, that is worrisome.

Reply:
Good point. Done.